crates/net/downloaders/src/file_client.rs

```diff
@@ -235,6 +246,94 @@ impl FileClient {
     }
 }
 
+struct BlockCertificateChecker {
+    certified_data: CertifiedResult<did::Block<did::H256>>,
+    evmc_principal: Principal,
+    ic_root_key: Vec<u8>
+}
+
+impl BlockCertificateChecker {
+    async fn create_certificate_checker(reqwest_client: &ReqwestClient, evcm_principal: String, ic_root_key: String) -> Result<Self, FileClientError> {
+        let evmc_principal = Principal::from_text(evcm_principal).map_err(|e| FileClientError::ProviderError(format!("failed to parse principal: {e}")))?;
+        let ic_root_key = hex::decode(&ic_root_key).map_err(|e| FileClientError::ProviderError(format!("failed to parse IC root key: {e}")))?;
+
+        let request = Request::Single(
+            jsonrpc_core::Call::MethodCall(
+                MethodCall{
+                    jsonrpc: Some(jsonrpc_core::Version::V2),
+                    method: "ic_getLastCertifiedBlock".to_string(),
+                    params: jsonrpc_core::Params::Array(vec![]),
+                    id: Id::Null,
+                }
+            )
+        );
+
+        let jsonrpc_core::Response::Single(output) =  reqwest_client.send_rpc_request(request).await.map_err(|err| FileClientError::ProviderError(format!("JSON RPC call failed: {err:?}")))? else {
+            return Err(FileClientError::ProviderError("invalid response, expected single".to_string()))
+        };
+
+        match output {
+            Output::Success(succ) => {
+                let certified_data: CertifiedResult<did::Block<did::H256>> = serde_json::from_value(succ.result).map_err(|err| FileClientError::ProviderError(format!("failed to deserialize certified block data: {err:?}")))?;
+                Ok(Self{certified_data, evmc_principal, ic_root_key})
+            },
+            Output::Failure(fail) => Err(FileClientError::ProviderError(format!("JSON RPC call returned error: {fail:?}"))),
+        }
+    }
+
+    fn get_block_number(&self) -> u64 {
+        self.certified_data.data.number.0.as_u64()
+    }
+
+    fn check_block(&self, block: &did::Block<did::Transaction>) -> Result<(), FileClientError> {
+        if block.number < self.certified_data.data.number {
+            return Ok(());
```

Comment:
why?

Reply:
Because we can't query the arbitrary certified block from the EVM canister, only the latest one. So fetch all blocks up to it